MobileCenterPush/MobileCenterPush/MSPush.m

```diff
@@ -0,0 +1,277 @@
+#import "MSPush.h"
+#import "MSPushLog.h"
+#import "MSPushPrivate.h"
+#import "MSPushInternal.h"
+#import "MSDeviceTracker.h"
+#import "MSMobileCenterInternal.h"
+
+#if defined(__IPHONE_10_0) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
+
+#import <UserNotifications/UserNotifications.h>
+
+#endif
+
+/**
+ * Service storage key name.
+ */
+static NSString *const kMSServiceName = @"Push";
+
+/**
+ * Key for storing push token
+ */
+static NSString *const kMSPushServiceStorageKey = @"kmspushservicepushstoringkey";
+
+/**
+ * Singleton
+ */
+static MSPush *sharedInstance = nil;
+static dispatch_once_t onceToken;
+
+@implementation MSPush
+
+@synthesize deviceTokenHasBeenSent;
+@synthesize isRequestInProgress;
+
+#pragma mark - Service initialization
+
+- (instancetype)init {
+
+  self = [super init];
+
+  if (self) {
+
+    deviceTokenHasBeenSent = NO;
+    isRequestInProgress = NO;
+  }
+
+  return self;
+}
+
+#pragma mark - MSServiceInternal
+
++ (instancetype)sharedInstance {
+  dispatch_once(&onceToken, ^{
+    if (sharedInstance == nil) {
+      sharedInstance = [self new];
+    }
+  });
+  return sharedInstance;
+}
+
+- (void)startWithLogManager:(id<MSLogManager>)logManager appSecret:(NSString *)appSecret {
+  [super startWithLogManager:logManager appSecret:appSecret];
+
+  MSLogVerbose([MSPush logTag], @"Started push service.");
+
+  if( [MSPush isEnabled] ) {
+
+    [self registerForRemoteNotifications];
+  }
+}
+
++ (NSString *)logTag {
+  return @"MobileCenterPush";
+}
+
+- (NSString *)storageKey {
+  return kMSServiceName;
+}
+
+- (MSPriority)priority {
+  return MSPriorityDefault;
+}
+
+- (MSInitializationPriority)initializationPriority {
+  return MSInitializationPriorityDefault;
+}
+
+#pragma mark - MSPush
+
++ (void)registerForRemoteNotifications {
+  [[self sharedInstance] registerForRemoteNotifications];
+}
+
++ (void)didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
+  [[self sharedInstance] didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
+}
+
++ (void)didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
+  [[self sharedInstance] didFailToRegisterForRemoteNotificationsWithError:error];
+}
+
+#pragma mark - MSServiceAbstract
+
+- (void)applyEnabledState:(BOOL)isEnabled {
+  [super applyEnabledState:isEnabled];
+
+  if (isEnabled) {
+
+    MSLogInfo([MSPush logTag], @"Push service has been enabled.");
+
+    if( !self.deviceTokenHasBeenSent ) {
+
+      [self registerForRemoteNotifications];
+    }
+  } else {
+
+    [[UIApplication sharedApplication] unregisterForRemoteNotifications];
+
+    MSLogInfo([MSPush logTag], @"Push service has been disabled.");
+  }
+}
+
+#pragma mark - Private methods
+
++ (void)resetSharedInstance {
+
+  // resets the once_token so dispatch_once will run again
+  onceToken = 0;
+  sharedInstance = nil;
+}
+
+- (void) registerForRemoteNotifications {
+
+  if( self.isRequestInProgress )
```

Comment:
This is breaking the design of the SDK. Seems like it will be able to send a request before start. What's the purpose of having static method of `registerForRemoteNotifications`?